### ISSUE TITLE
Create a metrics deprecation policy

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#33995]
During the 2024-07-24 community meeting, the topic was raised about how to change metrics over time.

One of the key questions here is whether the Cilium community wants to provide longer stability for metrics, such as deprecating it in one release while introducing a replacement, then only removing it in the subsequent release.

**Comment by joestringer** [comment]
One of the points that was raised in general with changing metrics is that if you are changing a metric, then it is good practice to create a new metric with a different name to minimize impact on users who are consuming the current metric in a different form.

**Comment by tommyp1ckles** [comment]
The approach I've taken in the past is:

## Add to Upgrade Notes and release with new Cilium Version:
* Changes that are unlikely to even require user changes, such as adding a new label, can simply be added to upgrade notes.
* Changes that may require intervention during upgrade, but don't fundamentally change the metric (i.e. like changing the name of a label).

## Create new metric and deprecate old one

* In cases where the old metric had fundamental problems, or was often misused it sometimes makes sense to add a replacement metric with a new name, and deprecate -> remove the old one on next_version + 1.
* Major changes, such as changing the metric type probably fall into here as well.

## Other Scenario: Create new metric, keep old one but maybe make it opt-in:

* Sometimes a reasonable decision might just be to disable the old metric by default, and add the new one one which is enabled by default.  This case might arise when the old metric is still useful, but has some major limitations or is difficult to use (i.e. something with high cardinality unsuitable for a large production environment, but still useful for debugging purposes).

**Comment by marseel** [comment]
I think long-term we want to have a different process. My 2 cents:
- Metrics maturity should be decoupled from feature maturity - You might want to add a new "beta" metric to an already existing stable feature and you don't want to go through the whole process of deprecation if you decide to change it later on
- Stable metrics - following deprecation policy etc
- We have some "developer" metrics that we don't expect users to use - we should be always able to change them + prevent users from enabling them - for example https://github.com/cilium/cilium/issues/27659

K8s has a similar maturity of metrics. 

Additionally, have some general guidelines around the cardinality of metrics so we don't introduce metrics that need to be disabled at X nodes / Y pods etc scale.

**Comment by joestringer** [comment]
Yep better developer documentation would be good. Currently we rely 100% on @cilium/metrics to provide this feedback (codeowner review scope below, [docs](https://docs.cilium.io/en/stable/contributing/development/reviewers_committers/review_process/#project-wide)), so it would be good for folks in that group to clarify the guidelines.

> [@cilium/metrics](https://github.com/orgs/cilium/teams/metrics): Provide recommendations about the types, names and labels for metrics to follow best practices. This includes considering the cardinality impact of metrics being added or extended.

**Comment by chancez** [comment]
Some considerations that don't fit as nicely into the things already discussed:
- Adding and removing labels. These impact existing metrics, adding new labels increases cardinality, removing them can break existing queries.
- Changing histogram bucket sizes. This can be ok, but it will cause a period where queries using `histogram_quantile` can see some interesting results.

I general, I think we should be looking to improve the fundamental architecture of how users interact and enable metrics. I'm not super fond of how we do it in hubble, though there's been some movement there that's been in the right direction. Concretely I would love for Cilium to adopt the following, which can help with most issues being mentioned:

- A _structured_ configuration (file) for enabling/disabling metrics.
- The configuration at a minimum would allow:
   - Enabling metrics that are disabled by default
   - Disabling metrics that are enabled by default
   - Configuring labels on metrics (some labels may be default enabled, some might not).
   - Configuring histogram bucket sizes
- Another consideration for many of our metrics is sometimes they apply to certain entities in the cluster.  Sometimes this can explode cardinality, or at the least, introduce more series than you wish to deal with, or pay for managing. One way to improve upon this situation is to be able to add filtering to metrics, defining which entities you actually do metrics accounting for, within a given metric.
  - For example: flow counts might be tied to a given endpoint or pod via some labels. Being able to selectively specify which namespaces, the metric accounting is done for is a really useful feature. For more details, see the RFC I link at the end.

This would enable more granular opt-in/opt-out to various aspects of our metrics, and potentially allows us to more easily determine if a user is using something that's been deprecated. I know we already have some of this, but it's not exactly _structured_, and at from my memory with how Cilium metrics are enabled/disabled, along with how you configure Hubble metrics are configured, we've basically invented a mini-DSL for configuring things and it's pretty painful to be honest, hence my focus on a structured config (file) for managing these things. 

Some references about where a contributor recently wanted to improve upon the Hubble Metrics cardinality. The goal was cardinality, but to enable them to actually do this, required more configuration options to filter the pods/namespaces the metrics accounting occurred on, which we didn't want to accept more ways to configure metrics with our current mini-DSL, so we direct them to a new configuration format to achieve it.

- https://github.com/cilium/design-cfps/blob/main/cilium/CFP-30788-managing-hubble-metrics-cardinality.md
- https://github.com/cilium/cilium/pull/32294

Lastly, I would encourage us to look at how envoy and Istio manage metrics. They do pretty much everything I've suggested, which gives the users flexibility where it's really needed. A one size fits all approach for metrics rarely works for base infrastructure components like Cilium. Your histogram buckets, labels, etc, may work for your environment, but not mine. Sometimes things need the option to be tuned to the particular environment. Here's some references on different ways you can do this in envoy and Istio:

- https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto
- https://istio.io/latest/docs/tasks/observability/metrics/customize-metrics/
- https://istio.io/latest/docs/tasks/observability/metrics/telemetry-api/

The Envoy/Istio approach may makes more sense when thinking about Hubble, but some ideas can still make sense for core Cilium metrics as well.

One thing I would discourage is introducing a new CRD for this. In my opinion, a config file the agent reads is a much better, less complicated starting point, and restricts configuration to administrators. In the future, a CRD could be built on top of these config files, and/or managed by a "Cilium Install/LIfeCycle Operator" or similar, where the CRD merely acts as a "DSL" that will result in said operator re-generating the underlying config file proposed. This layering works nicely, and has been used successfully in other CNCF projects like Prometheus (eg: ServiceMonitor's just end up being translated to some scrape rules in the Prometheus config file). Additionally, by not having a CRD as the core primitive for configuring these things, it makes it easier to use when running Cilium outside of Kubernetes.

**Comment by tommyp1ckles** [comment]
I like the idea of having better configuration for metrics

> One thing I would discourage is introducing a new CRD for this. In my opinion, a config file the agent reads is a much better, less complicated starting point, and restricts configuration to administrators.

+1

What about just having a section in the values value, that would translate into

**Comment by tommyp1ckles** [comment]
> Metrics maturity should be decoupled from feature maturity - You might want to add a new "beta" metric to an already existing stable feature and you don't want to go through the whole process of deprecation if you decide to change it later on

Yeah I like this, even thinking about what metrics are stable at this point vs ones that we may end up changing would be useful for both someone working on them as well as users.

> K8s has a similar maturity of metrics - https://kubernetes.io/docs/reference/instrumentation/metrics/

We could probably easily extend our own cell.Metric types to include a stability/maturity field.

**Comment by chancez** [comment]
> What about just having a section in the values value, that would translate into

That's fine. The Hubble Metrics CFP we directed them to making it possible to configure within the helm values, or let the user supply their own configMap so they can manage it outside of the helm install . This gives flexibility for if a user wants to automate management of the configMap outside of helm (eg: with a custom operator, or similar), which could make it possible for non-cluster-admins to administer these options in some way. This is particularly valuable to K8s-as-a-service vendors which might manage the metrics configuration on behalf of users through eg: their UI.
